internal/pkg/cli/deploy/deploy.go

```diff
@@ -0,0 +1,850 @@
+type WorkspaceReader interface {
+	manifestReader
+	pathFinder
+}
+
+// Interpolator interpolates a manifest.
+type Interpolator interface {
+	Interpolate(s string) (string, error)
+}
+
+// progress is the interface to inform the user that a long operation is taking place.
+type Progress interface {
+	Start(label string)
+	Stop(label string)
+}
+
+type manifestReader interface {
+	ReadWorkloadManifest(name string) (workspace.WorkloadManifest, error)
+}
+
+type pathFinder interface {
+	Path() (string, error)
+}
+
+type fileReader interface {
+	ReadFile(string) ([]byte, error)
+}
+
+type workloadDeployer struct {
+	name          string
+	app           *config.Application
+	env           *config.Environment
+	imageTag      string
+	s3Bucket      string
+	mft           interface{}
+	workspacePath string
+}
+
+// NewWorkloadDeployerInput is the input to for workloadDeployer constructor.
+type NewWorkloadDeployerInput struct {
+	Name     string
+	App      *config.Application
+	Env      *config.Environment
+	ImageTag string
+	S3Bucket string
+}
+
+// NewWorkloadDeployer is the constructor for workloadDeployer.
+func NewWorkloadDeployer(in *NewWorkloadDeployerInput) (*workloadDeployer, error) {
+	ws, err := workspace.New()
+	if err != nil {
+		return nil, fmt.Errorf("new workspace: %w", err)
+	}
+	mft, err := workloadManifest(&workloadManifestInput{
+		name:         in.Name,
+		appName:      in.App.Name,
+		envName:      in.Env.Name,
+		interpolator: manifest.NewInterpolator(in.App.Name, in.Env.Name),
+		ws:           ws,
+		unmarshal:    manifest.UnmarshalWorkload,
+	})
+	if err != nil {
+		return nil, err
+	}
+	workspacePath, err := ws.Path()
+	if err != nil {
+		return nil, fmt.Errorf("get workspace path: %w", err)
+	}
+	return &workloadDeployer{
+		name:          in.Name,
+		app:           in.App,
+		env:           in.Env,
+		imageTag:      in.ImageTag,
+		s3Bucket:      in.S3Bucket,
+		mft:           mft,
+		workspacePath: workspacePath,
+	}, nil
+}
+
+// UploadArtifactsInput is the input of UploadArtifacts.
+type UploadArtifactsInput struct {
+	FS                 fileReader
+	Uploader           Uploader
+	Templater          Templater
+	ImageBuilderPusher ImageBuilderPusher
+}
+
+// UploadArtifactsOutput is the output of UploadArtifacts.
+type UploadArtifactsOutput struct {
+	ImageDigest string
+	EnvFileARN  string
+	AddonsURL   string
+}
+
+// DeployWorkloadInput is the input of DeployWorkload.
+type DeployWorkloadInput struct {
+	ImageDigest    string
+	EnvFileARN     string
+	AddonsURL      string
+	RootUserARN    string
+	AppTags        map[string]string
+	ResourceTags   map[string]string
+	ImageRepoURL   string
+	ForceNewUpdate bool
+
+	CustomResourceUploader CustomResourcesUploader
+	S3Uploader             Uploader
+	SNSTopicsLister        SNSTopicsLister
+	ServiceDeployer        ServiceDeployer
+	NewSvcUpdater          func(func(*session.Session) ServiceForceUpdater)
+	NewAppVersionGetter    func(string) (VersionGetter, error)
```

Comment:
Is there a reason why we cant do this instead:
```suggestion
	ForceUpdater          ServiceForceUpdater
	AppVersionGetter      VersionGetter
```

Reply:
For `NewSvcUpdater` because we won't know what updater we should initialize.